go/cmd/dolt/commands/push.go

```diff
@@ -85,95 +95,236 @@ func (cmd PushCmd) Exec(ctx context.Context, commandStr string, args []string, d
-					"To have this happen automatically for branches without a tracking\n" +
-					"upstream, see 'push.autoSetupRemote' in 'dolt config --help'.").Build()
+				return fmt.Errorf("failed to get tracking ref info: %w", err)
 			}
 
-		case env.ErrInvalidSetUpstreamArgs:
-			verr = errhand.BuildDError("error: --set-upstream requires <remote> and <refspec> params.").SetPrintUsage().Build()
-		default:
-			verr = errhand.VerboseErrorFromError(err)
+			cli.Printf("To %s\n", targetRemote.Url)
+			cli.Printf("! [rejected]          %s -> %s (non-fast-forward)\n", srcRef, destRef)
+			cli.Printf("error: failed to push some refs to '%s'\n", targetRemote.Url)
+			cli.Println("hint: Updates were rejected because the tip of your current branch is behind")
+			cli.Println("hint: its remote counterpart. Integrate the remote changes (e.g.")
+			cli.Println("hint: 'dolt pull ...') before pushing again.")
 		}
-		return HandleVErrAndExitCode(verr, usage)
+		return err
 	}
 
-	remoteDB, err := opts.Remote.GetRemoteDB(ctx, dEnv.DoltDB.ValueReadWriter().Format(), dEnv)
+	postPushRemoteBranches, err := getRemoteBranches(queryist, sqlCtx)
 	if err != nil {
-		err = actions.HandleInitRemoteStorageClientErr(opts.Remote.Name, opts.Remote.Url, err)
-		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+		return fmt.Errorf("failed to get post-push remote branches: %w", err)
 	}
 
-	tmpDir, err := dEnv.TempTableFilesDir()
-	if err != nil {
-		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	changesMade := getChangesMade(remoteBranches, postPushRemoteBranches)
+	if !changesMade {
+		cli.Println("Everything up-to-date")
 	}
-	var verr errhand.VerboseError
-	err = actions.DoPush(ctx, dEnv.RepoStateReader(), dEnv.RepoStateWriter(), dEnv.DoltDB, remoteDB, tmpDir, opts, buildProgStarter(defaultLanguage), stopProgFuncs)
+
+	return nil
+}
+
+func getChangesMade(remoteBranches map[string]remoteBranchInfo, postPushRemoteBranches map[string]remoteBranchInfo) bool {
+	changesMade := false
+	if len(remoteBranches) != len(postPushRemoteBranches) {
+		changesMade = true
+	} else {
+		for name, remoteBranch := range remoteBranches {
+			updatedRemoteBranch, ok := postPushRemoteBranches[name]
+			if !ok {
+				changesMade = true
+				break
+			}
+			if remoteBranch.Hash != updatedRemoteBranch.Hash {
+				changesMade = true
+				break
+			}
+		}
+	}
+	return changesMade
+}
+
+func getRemotes(queryist cli.Queryist, sqlCtx *sql.Context) (map[string]remoteInfo, error) {
+	rows, err := GetRowsForSql(queryist, sqlCtx, "select * from dolt_remotes")
 	if err != nil {
-		verr = printInfoForPushError(err, opts.Remote, opts.DestRef, opts.RemoteRef)
+		return nil, fmt.Errorf("failed to read dolt remotes: %w", err)
 	}
 
-	if opts.SetUpstream {
-		err := dEnv.RepoState.Save(dEnv.FS)
+	remotes := map[string]remoteInfo{}
+	for _, row := range rows {
+		name := row[0].(string)
+		url := row[1].(string)
```

Comment:
Pull out all columns here, that lets the reader see clearly the full structure of the response table. Don't use row[n] anywhere below.

Reply:
Restructuring now.